pop-api/examples/fungibles/utils.rs

```diff
@@ -0,0 +1,196 @@
+// PSP22Metadata functions.
+pub const TOKEN_NAME: &str = "Psp22Metadata::token_name";
+pub const TOKEN_SYMBOL: &str = "Psp22Metadata::token_symbol";
+pub const TOKEN_DECIMALS: &str = "Psp22Metadata::token_decimals";
+// PSP22Mintable functions.
+pub const MINT: &str = "Psp22Mintable::mint";
+// PSP22Burnable functions.
+pub const BURN: &str = "Psp22Burnable::burn";
+
+/// This is used to resolve type mismatches between the `AccountId` in the quasi testing environment and the
+/// contract environment.
+pub(super) fn account_id_from_slice(s: &[u8; 32]) -> AccountId {
+	AccountId::decode(&mut &s[..]).expect("Should be decoded to AccountId")
+}
+
+/// Get the last event from pallet contracts.
+pub(super) fn last_contract_event(session: &Session<Sandbox>) -> Option<Vec<u8>> {
+	session.record().last_event_batch().contract_events().last().cloned()
+}
+
+/// Execute a contract method and exepct CallReverted error to be returned.
+pub(super) fn expect_call_reverted(
+	session: &mut Session<Sandbox>,
+	function: &str,
+	params: Vec<String>,
+	err: PSP22Error,
+) {
+	let call = session.call::<String, ()>(function, &params, None);
+	if let Err(SessionError::CallReverted(error)) = call {
+		assert_eq!(error[1..], Err::<(), PSP22Error>(err).encode());
+	}
+}
+
+// Call a contract method and decode the returned data.
+pub(super) fn decoded_call<T: Decode>(
+	session: &mut Session<Sandbox>,
+	func_name: &str,
+	input: Vec<String>,
+	endowment: Option<Balance>,
+) -> Result<T, Box<dyn std::error::Error>> {
+	session.call(func_name, &input, endowment)??;
+	Ok(session.record().last_call_return_decoded::<T>()??)
+}
+
+// Check if the event emitted correctly.
+pub(super) fn assert_event(session: &mut Session<Sandbox>, event: Vec<u8>) {
+	let contract_events = session.record().last_event_batch().contract_events();
+	let last_event = contract_events.last().unwrap().to_vec();
+	assert_eq!(last_event, event.as_slice());
+}
+
+// Test methods for deployment with constructor function.
+
+pub(super) fn deploy_with_new_constructor(
+	session: &mut Session<Sandbox>,
+	bundle: ContractBundle,
+	id: TokenId,
+	min_balance: Balance,
+) -> Result<AccountId32, SessionError> {
+	session.deploy_bundle(
+		bundle,
+		"new",
+		&[id.to_string(), min_balance.to_string()],
+		NO_SALT,
+		Some(INIT_VALUE),
+	)
+}
+
+pub(super) fn deploy_with_new_existing_constructor(
+	session: &mut Session<Sandbox>,
+	bundle: ContractBundle,
+	id: TokenId,
+) -> Result<AccountId32, SessionError> {
+	session.deploy_bundle(bundle, "new_existing", &[id.to_string()], NO_SALT, Some(INIT_VALUE))
+}
+
+// Test methods for `PSP22`.
+
+pub(super) fn total_supply(session: &mut Session<Sandbox>) -> Balance {
+	decoded_call::<Balance>(session, TOTAL_SUPPLY, vec![], None).unwrap()
+}
+
+pub(super) fn balance_of(session: &mut Session<Sandbox>, owner: AccountId32) -> Balance {
+	decoded_call::<Balance>(session, BALANCE_OF, vec![owner.to_string()], None).unwrap()
```

Comment:
Why do we have to pass the parameters as string and not encoded? The latter would be better as it mimics how it works in reality.

Reply:
Because `session.call()` method from `pop-drink` asks for `Vec<String>`.